src/tools/pkcs11-tool.c

```diff
@@ -3545,57 +3545,75 @@ unwrap_key(CK_SESSION_HANDLE session)
+	} else if (strncasecmp(opt_key_type, "GENERIC:", strlen("GENERIC:")) == 0) {
+		length = opt_key_type + strlen("GENERIC:");
+		key_type = CKK_GENERIC_SECRET;
+	} else if (strncasecmp(opt_key_type, "RSA:", strlen("RSA:")) == 0) {
+		length = "0"; // No key length for RSA keys
+		key_type = CKK_RSA;
+		class = CKO_PRIVATE_KEY;
+	} else if (strncasecmp(opt_key_type, "EC:", strlen("EC:")) == 0) {
+		length = "0"; // No key length for EC keys
+		key_type = CKK_EC;
+		class = CKO_PRIVATE_KEY;
+	} else {
+		util_fatal("Unsupported key type %s", opt_key_type);
+	}
 
-		FILL_ATTR(keyTemplate[n_attr], CKA_KEY_TYPE, &key_type, sizeof(key_type));
-		n_attr++;
+	FILL_ATTR(keyTemplate[n_attr], CKA_KEY_TYPE, &key_type, sizeof(key_type));
+	n_attr++;
 
-		if (opt_is_sensitive != 0) {
-			FILL_ATTR(keyTemplate[n_attr], CKA_SENSITIVE, &_true, sizeof(_true));
-			n_attr++;
-		} else {
-			FILL_ATTR(keyTemplate[n_attr], CKA_SENSITIVE, &_false, sizeof(_false));
-			n_attr++;
-		}
+	if (opt_is_sensitive != 0) {
+		FILL_ATTR(keyTemplate[n_attr], CKA_SENSITIVE, &_true, sizeof(_true));
+	} else {
+		FILL_ATTR(keyTemplate[n_attr], CKA_SENSITIVE, &_false, sizeof(_false));
+	}
+	n_attr++;
 
-		if (opt_key_usage_default || opt_key_usage_decrypt) {
+	if (opt_key_usage_default || opt_key_usage_decrypt) {
+		if (class != CKO_PRIVATE_KEY) {
 			FILL_ATTR(keyTemplate[n_attr], CKA_ENCRYPT, &_true, sizeof(_true));
 			n_attr++;
-			FILL_ATTR(keyTemplate[n_attr], CKA_DECRYPT, &_true, sizeof(_true));
-			n_attr++;
 		}
-		if (opt_key_usage_wrap) {
+		FILL_ATTR(keyTemplate[n_attr], CKA_DECRYPT, &_true, sizeof(_true));
+		n_attr++;
+	}
+	if (opt_key_usage_wrap) {
+		if (class != CKO_PRIVATE_KEY) {
 			FILL_ATTR(keyTemplate[n_attr], CKA_WRAP, &_true, sizeof(_true));
 			n_attr++;
-			FILL_ATTR(keyTemplate[n_attr], CKA_UNWRAP, &_true, sizeof(_true));
-			n_attr++;
-		}
-
-		if (opt_is_extractable != 0) {
-			FILL_ATTR(keyTemplate[n_attr], CKA_EXTRACTABLE, &_true, sizeof(_true));
-			n_attr++;
-		} else {
-			FILL_ATTR(keyTemplate[n_attr], CKA_EXTRACTABLE, &_false, sizeof(_true));
-			n_attr++;
 		}
-		/* softhsm2 does not allow to attribute CKA_VALUE_LEN, but MyEID card must have this attribute
-                   specified. We set CKA_VALUE_LEN only if the user sets it in the key specification. */
-		key_length = (unsigned long)atol(length);
-		if (key_length != 0) {
-			FILL_ATTR(keyTemplate[n_attr], CKA_VALUE_LEN, &key_length, sizeof(key_length));
+		FILL_ATTR(keyTemplate[n_attr], CKA_UNWRAP, &_true, sizeof(_true));
+		n_attr++;
+	}
+	if (opt_key_usage_sign) {
+		if (class != CKO_PRIVATE_KEY) {
+			FILL_ATTR(keyTemplate[n_attr], CKA_VERIFY, &_true, sizeof(_true));
 			n_attr++;
 		}
+		FILL_ATTR(keyTemplate[n_attr], CKA_SIGN, &_true, sizeof(_true));
+		n_attr++;
+	}
+
+	if (opt_is_extractable != 0) {
+		FILL_ATTR(keyTemplate[n_attr], CKA_EXTRACTABLE, &_true, sizeof(_true));
+	} else {
+		FILL_ATTR(keyTemplate[n_attr], CKA_EXTRACTABLE, &_false, sizeof(_true));
```

Comment:
technically, this should be `sizeof(_false)`, but given the size is same, it does not matter. It might be just confusing for a reader or less clever static analyzers ...